mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/hook/ClientEventHookManagerImpl.java

```diff
@@ -0,0 +1,170 @@
+                        if (eventMessage != null) {
+                            eventMessages.add(eventMessage);
+                        }
+                    }
+                    if (eventMessages.isEmpty()) {
+                        continue;
+                    }
+
+                    // covert to mqtt message
+                    List<MqttPublishMessage> eventPublishMessages = toMqttMessage(eventMessages);
+                    CompletableFuture<HookResult> eventHookResult = clientEventHook.doHook(eventPublishMessages);
+                    Preconditions.checkNotNull(eventHookResult, "Put client events to LMQ error by Null hook result.");
+
+                    eventHookResult.whenComplete((hookResult, throwable) -> {
+                        if (throwable != null) {
+                            logger.error("Put client events to LMQ error:", throwable);
+                        }
+                        if (logger.isDebugEnabled()) {
+                            logger.debug("Put client events to LMQ success: {}", hookResult);
+                        }
+                    });
+                } catch (Throwable t) {
+                    logger.error("ClientEventHookExecution error: ", t);
+                } finally {
+                    // release msgId
+                    releaseMsgId(eventMessages);
+                }
+            }
+        }
+    }
+
+    public List<MqttPublishMessage> toMqttMessage(List<ClientEventMessage> eventMessages) {
+        List<MqttPublishMessage> eventPublishMessages = new ArrayList<>(eventMessages.size());
+
+        for (ClientEventMessage eventMessage : eventMessages) {
+            MqttFixedHeader mqttFixedHeader = new MqttFixedHeader(MqttMessageType.PUBLISH,
+                    false, MqttQoS.AT_LEAST_ONCE, false, 0);
+            MqttPublishVariableHeader mqttPublishVariableHeader = new MqttPublishVariableHeader(
+                    CLIENT_EVENT_ORIGIN_TOPIC, eventMessage.getPacketId());
+            ByteBuf payload = Unpooled.wrappedBuffer(eventMessage.toString().getBytes(StandardCharsets.UTF_8));
+            MqttPublishMessage eventPublishMessage = new MqttPublishMessage(mqttFixedHeader, mqttPublishVariableHeader, payload);
+
+            eventPublishMessages.add(eventPublishMessage);
+        }
+        return eventPublishMessages;
+    }
+
+    private void releaseMsgId(List<ClientEventMessage> eventMessages) {
+        for (ClientEventMessage eventMessage : eventMessages) {
+            mqttMsgId.releaseId(eventMessage.getPacketId(), eventMessage.getClientId());
+        }
+    }
+
+    @Override
+    public void putClientEvent(Channel channel, ClientEventType eventType) {
+        ClientEventMessage eventMessage = new ClientEventMessage(eventType);
+        eventMessage.setChannelId(ChannelInfo.getId(channel))
+                .setClientId(ChannelInfo.getClientId(channel))
+                .setPacketId(mqttMsgId.nextId(ChannelInfo.getClientId(channel)));
```

Comment:
The id here can always be -1，because mqtt consumers don't care about this id

Reply:
Thx. I would keep this.